url2114d537a2_ForNeVeR_lanterna_tgz-pJ8-com_googlecode_lanterna_test_gui_DifferentKindsOfListBoxesTestJ8/src/com/googlecode/lanterna/terminal/text/UnixTerminal.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * This file is part of lanterna (http://code.google.com/p/lanterna/).
+ * 
+ * lanterna is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ * Copyright (C) 2010-2012 Martin
+ */
+
+package com.googlecode.lanterna.terminal.text;
+
+import com.googlecode.lanterna.LanternaException;
+import com.googlecode.lanterna.input.GnomeTerminalProfile;
+import com.googlecode.lanterna.input.PuttyProfile;
+import com.googlecode.lanterna.terminal.TerminalSize;
+import java.io.*;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.nio.charset.Charset;
+
+/**
+ * A common ANSI terminal extention with support for Unix resize signals and 
+ * the stty program to control cbreak and key echo
+ * @author Martin
+ */
+public class UnixTerminal extends ANSITerminal
+{
+    private final UnixTerminalSizeQuerier terminalSizeQuerier;
+            
+    public UnixTerminal(
+            InputStream terminalInput, 
+            OutputStream terminalOutput, 
+            Charset terminalCharset)
+    {
+        this(terminalInput, terminalOutput, terminalCharset, null);
+    }
+            
+    public UnixTerminal(
+            InputStream terminalInput, 
+            OutputStream terminalOutput, 
+            Charset terminalCharset,
+            UnixTerminalSizeQuerier customSizeQuerier)
+    {
+        super(terminalInput, terminalOutput, terminalCharset);
+        this.terminalSizeQuerier = customSizeQuerier;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;customSizeQuerier&#39; (index: 3 | type: UnixTerminalSizeQuerier). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;terminalInput&#39;, &#39;terminalCharset&#39;, &#39;terminalOutput&#39;. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).